[vcpkg] User help: selecting-library-features.md (#14189)
microsoft/vcpkg: docs/users/selecting-library-features.md

# Selecting library features

## Installing a library

> By now you have tried out installing some library with `vcpkg`. We will look at [llvm](https://llvm.org/) as an example.

Reviewer: I don't understand the purpose of this sentence.

Author: If the user has come all the way to read this manual page, definitely they have "some" experience with vcpkg.

LLVM is a compiler infrasture. It supports optimizing llvm-ir and generating machine code.
You could install it using:

```powershell

> .\vcpkg install llvm

Reviewer: ```suggestion
> .\vcpkg install llvm
```

```

> If you have 64bit Windows installed, and if you are interested in only x64 libraries, you can do:

Reviewer: ```suggestion
On Windows, this will install the 32-bit x86 LLVM, since that's the default triplet on Windows.
If you are building for 64-bit Windows instead, you can use the following command to change the default triplet:
```

```powershell

> .\vcpkg install --triplet x64-windows-static llvm

Reviewer: ```suggestion
> .\vcpkg install --triplet x64-windows llvm
```

```

> You can read more about [triplets](https://vcpkg.readthedocs.io/en/latest/users/triplets/) here.

Reviewer: ```suggestion
We have more documentation on triplets [here](triplets.md).
```

> Sadly currently we can't choose build type `debug` or `release` using command line switches.

Reviewer: ```suggestion
```

We build both debug and release for a reason. This line shouldn't be here as it's not unfortunate.

Author: But there is no way to build just one or another. What's the reason to force both?

Reviewer: Because you generally need both; if you want to RFC a feature for doing this, that's fine, but it shouldn't be "unfortunate" in our docs.

> With llvm installed, if we try to execute:

Reviewer: ```suggestion
With llvm now installed, we can execute:
```

```powershell

> # llc takes llvm IR and generates machine code
> .\packages\llvm_x64-windows-static\bin\llc.exe --version

Reviewer: ```suggestion
> # llc takes llvm IR and generates machine code
> .\installed\x86-windows\bin\llc.exe --version # or x86-windows, or replace with the actual triplet
```

```

we see:

```powershell
  Registered Targets:
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64
```

## Installing additional features

> But [llvm (10) supports many more targets](https://llvm.org/docs/GettingStarted.html#local-llvm-configuration), `AArch64, AMDGPU, ARM, BPF, Hexagon, Mips, MSP430, NVPTX, PowerPC, Sparc, SystemZ, X86, XCore`.

Reviewer: ```suggestion
But [llvm supports many more targets](https://llvm.org/docs/GettingStarted.html#local-llvm-configuration), from ARM to SPARC to SystemZ.
```

> How can we have arm target?

Reviewer: ```suggestion
However, clearly our current installation doesn't include ARM as a target;
thus, we need to learn how vcpkg allows us to install other LLVM targets.
The llvm port allows this via the "target-*" features.
```

If we do:

```powershell
.\vcpkg search llvm
```

We can see:

```
llvm                 10.0.0#6         The LLVM Compiler Infrastructure
llvm[clang]                           Build C Language Family Front-end.
llvm[clang-tools-extra]               Build Clang tools.
...
llvm[target-all]                      Build with all backends.
llvm[target-amdgpu]                   Build with AMDGPU backend.
llvm[target-arm]                      Build with ARM backend.
...
```

> We can install "feature":

Reviewer: ```suggestion
We can install any of these targets by using the install-feature syntax:
```

```powershell

> .\vcpkg install --triplet x64-windows-static llvm[target-arm]

Reviewer: ```suggestion
> .\vcpkg install llvm[target-arm] # Installs LLVM with the ARM target
```

```

## Opting out of default feature

> When we installed llvm, we also got [clang](https://clang.llvm.org/) `.\packages\llvm_x64-windows-static\bin\clang.exe`. But what if we are not interested pre-bundled `clang`

Reviewer: ```suggestion
The llvm port includes a few default features that you as a user may not want: for example,
the `clang` feature is default, which means that `vcpkg install llvm` will also build and install clang.
If you are writing a compiler that uses LLVM as a backend,
you're likely not interested in installing clang as well,
and we can do that by disabling default features with the special `core` "feature":
```

> "c" language frontend (to convert c/c++ to llvm-ir)?
> We can only install llvm as:

Reviewer: ```suggestion

```

```powershell

> .\vcpkg install --triplet x64-windows-static llvm[core]

Reviewer: ```suggestion
> .\vcpkg install llvm[core,default-targets] # removing the default-feature with "core" also removes all of the default targets you get
```

```

> and later add targets and clang as desired.
> 

Reviewer: ```suggestion
```

# Further reading

> - [Feature Packages](specifications/feature-packages.md)

Reviewer: ```suggestion
- The [Feature Packages](specifications/feature-packages.md) specification was the initial design for features.
```
